server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,361 @@
+     * are tracked, not the overhead of the Set itself.
+     */
+    public long getSizeInBytes() {
+        long bytes = 0;
+        if (hashes != null) {
+            bytes = (hashes.size() * 16) + 8 + 4 + 1;
+        }
+        if (filters != null) {
+            bytes += filters.stream().mapToLong(CuckooFilter::getSizeInBytes).sum();
+        }
+        return bytes;
+    }
+
+
+    /**
+     * Merge `other` cuckoo filter into this cuckoo.  After merging, this filter's state will
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            if (hashes.size() > threshold) {
+                convert();
+            }
+        } else if (isSetMode && other.isSetMode == false) {
+            // Other is in cuckoo mode, so we convert our set to a cuckoo then merge collections.
+            // We could probably get fancy and keep our side in set-mode, but simpler to just convert
+            convert();
+            filters.addAll(other.filters);
+        } else if (isSetMode == false && other.isSetMode) {
+            // Rather than converting the other to a cuckoo first, we can just
+            // replay the values directly into our filter.
+            other.hashes.forEach(this::add);
+        } else {
+            // Both are in cuckoo mode, merge raw fingerprints
+
+            int current = 0;
+            CuckooFilter currentFilter = filters.get(current);
+
+            for (CuckooFilter filter : other.filters) {
+
+                // The iterator returns an array of longs corresponding to the
+                // fingerprints for buckets at the current position
+                Iterator<long[]> iter = filter.getBuckets();
+                int bucket = 0;
+                while (iter.hasNext()) {
+                    long[] fingerprints = iter.next();
+
+                    // We check to see if the fingerprint is present in any of the existing filters
+                    // (in the same bucket/alternate bucket), or if the fingerprint is empty.  In these cases
+                    // we can skip the fingerprint
+                    for (long fingerprint : fingerprints) {
+                        if (fingerprint == CuckooFilter.EMPTY || mightContainFingerprint(bucket, (int) fingerprint)) {
+                            continue;
+                        }
+                        boolean success = false;
+
+                        // If the fingerprint is new, we try to merge it into the filter at our `current` pointer.
+                        // This might fail (e.g. the filter is full), so we may have to try multiple times
```

Comment:
Since we have already discounted the case where this fingerprint already exists in one of the filters, and by virtue of the fact there are multiple filters all but the last filter must have been declared "full" before, won't the subsequent while loop likely always descend to the last filter? If so, shouldn't we simplify this by either starting `current` at `filters.size() - 1` or removing this logic and just keeping a reference to the last filter, updating this reference as we need to create more filters?

Reply:
++ I believe you're correct, good catch.  This is leftover from an earlier iteration that didn't check if the fingerprint existed.  It's not necessary now.